OpenRA.Game/TraitDictionary.cs

```diff
@@ -43,23 +43,72 @@ class TraitDictionary
 		static readonly Func<Type, ITraitContainer> CreateTraitContainer = t =>
 			(ITraitContainer)typeof(TraitContainer<>).MakeGenericType(t).GetConstructor(Type.EmptyTypes).Invoke(null);
 
-		readonly Dictionary<Type, ITraitContainer> traits = new Dictionary<Type, ITraitContainer>();
+		private ITraitContainer[] traits = null;
+
+		public TraitDictionary() {
+			Resize();
+		}
+
+		private void Resize() {
+			// New trait types have been registered. Ensure our array is same size as trait type index map.
+			int newSize = TraitTypeIndexMap.GetCount() + 64;
+			if (traits == null) {
```

Comment:
There are still a few (minor) style issues left (https://github.com/OpenRA/OpenRA/wiki/Coding-Standard):
Braces should be on their own line, and you should omit them around one-line statements.
As example, I'd suggest changing this to the following:
```
var newSize = TraitTypeIndexMap.GetCount() + 64;
if (traits == null)
	traits = new ITraitContainer[newSize];
else if (traits.Length < newSize)
	Array.Resize(ref traits, newSize);
```